### DISCUSSION TITLE
[Feature Request] Json Serialization:Azure IOT Hub SDK - System.Text.Json

### DISCUSSION THREAD

**Discussion by chandra-rayaprol** [Azure/azure-iot-sdk-csharp#2303] [General]
I would like to know whether 'System.Text.Json' is supported by the latest Azure IoT SDKs? To my understanding, Azure IOT SDK has 'Newtonsoft.json' as a dependency and is the default serializer used for device twins.

The default JsonSerializer for [Asp.net](http://asp.net/) core 6.0 is 'System.Text.Json'. It will be great if the same serializer can be used only across all azure SDK dependencies. Please advise.

**Comment by abhipsaMisra** [comment]
The latest preview of the Azure IoT .NET SDK allows you to specify your own serializer that would be used for serialization/deserialization of payloads exchanged with IoT hub - https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/iothub/device/src/Serialization/PayloadConvention.cs

You can try out the SDK and let us know your feedback: https://github.com/Azure/azure-iot-sdk-csharp/releases/tag/2022-11-03

Please note that this being a major version revision, there are breaking changes. This [migration guide](https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/SDK%20v2%20migration%20guide.md) can help you understand those change and guide you on how to port your code over to the v2 SDK.
If you have any further questions or encounter any issues, please let us know.

**Comment by chandra-rayaprol** [comment]
I was just working on migrating to the V2 library and encountered a few compilation errors.
1.'ClientTwinProperties.'ToJson()' is not defined. 
2. The class MessageSystemPropertyNames is marked as internal and inaccessible.
3. IQuery Interface is not found .  
Could you please help in resolving these errors.

**Reply by brycewang-microsoft** [reply]
These are caused by the new changes we brought in our v2 SDK. 

> 1. 'ClientTwinProperties.'ToJson()' is not defined.

We removed the "ToJson()" methods since [here](https://github.com/Azure/azure-iot-sdk-csharp/pull/2892). To get `ClientTwinProperties` as a JSON string, you can call `System.Text.Json.JsonSerializer.Serialize(clientTwinProperties)` instead.

> 2. The class MessageSystemPropertyNames is marked as internal and inaccessible.

Each of the property names in this class now has a proper setter/getter in the [Message](https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/iothub/service/src/Messaging/Models/Message.cs) class itself, like [here](https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/iothub/service/src/Messaging/Models/Message.cs#L42-L46). 

> 3. IQuery Interface is not found.

This interface was removed in v2. Instead, to execute a query and get the results, now you can call this method [here](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.devices.queryclient.createasync?view=azure-dotnet-preview#microsoft-azure-devices-queryclient-createasync-1(system-string-microsoft-azure-devices-queryoptions-system-threading-cancellationtoken)). More details can be found in the v2 migration [guide](https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/SDK%20v2%20migration%20guide.md#registrymanager).

**Comment by chandra-rayaprol** [comment]
It appears that the "System.Text.Json.JsonSerializer.Serialize" does not work for ClientTwinProperties. The code below produces a JSON string with no key values. `"Value":[]},{"Key":"id","Value":[]}].
`
var collection = new ClientTwinProperties();
        collection["Name"] = "TestValue";
        collection["id"] = 1116;
        var json = System.Text.Json.JsonSerializer.Serialize(collection);
        Console.WriteLine(json);
`

**Reply by abhipsaMisra** [reply]
We'll check this out @Raysapps, thanks.

**Reply by abhipsaMisra** [reply]
@Raysapps Looks like you were mistakenly guided towards the incorrect method, my apologies for that.
You'll want to use `string json = collection.GetPropertiesAsJson();` in order to retrieve the twin properties json.

Note that this method will return all of the twin's properties, but not the metadata. The version and metadata can be retrieved through their dedicated accessors.

I hope this helps.

**Comment by jonekdahl** [comment]
> If you have any further questions or encounter any issues, please let us know.

I'm also interested in switching the IotHub serialization to System.Text.Json but I need to configure the JsonSerializerOptions. I can't see a way to do that in the current client code (using 2.0.0-preview007). Is it possible or do I need to create my own PayloadConvention implementation to do that?

**Reply by jonekdahl** [reply]
Ping @abhipsaMisra since you were the one asking for more feedback.

**Reply by abhipsaMisra** [reply]
You're right. Our current implementation doesn't allow for a way to pass in the JsonSerializerOptions to the Serialize/Deserialize methods. Unfortunately, you'll need to create your own PayloadConvention implementation that does accept these params.

@andyk-ms FYI, to give any input about the future of the preview packages.
